### PR TITLE
Enable scatter e2e test cases

### PR DESCRIPTION
**Pull request by avery-blanchard**
Thanks for sending a pull request! Please make sure you read the [contributing guidelines](https://github.com/torch-spyre/torch-spyre/blob/main/CONTRIBUTING.md).

#### What type of PR is this?

- [ ] bug
- [x] feature
- [ ] documentation
- [x] cleanup

#### What this PR does:
Add scatter e2e tests for index_copy, index_put (including P=1 variant, second shape, two-target combined, and multi-dim scatter on dim1). Fix index initialization: _row_store uses arange, _full_index_store uses randperm with contiguous() before device transfer (DCI error workaround) for unique indices.

### CODE REVIEW

File: tests/inductor/test_indirect_access_scatter.py [hunk 1]
```diff
@@ -153,6 +164,129 @@ def kernel(out, src, idx):
...
+        return cache_dev, src_dev, idx_dev
+
+    def _assert_compiled_matches_cpu(self, kernel, *dev_args):
+        """Run `kernel` through torch.compile on device and require the result
+        to match the CPU eager reference -- a hard assertion, not an xfail."""
+        result = torch.compile(kernel, dynamic=False)(*dev_args)
+        reference = kernel(
+            *[a.cpu() if isinstance(a, torch.Tensor) else a for a in dev_args]
+        )
+        torch.testing.assert_close(result.cpu(), reference)
```
Comment (ani300) [inline tests/inductor/test_indirect_access_scatter.py:209]: The reference is computed from the device args *after* the compiled kernel has run. For the in-place `y[idx] = src` tests this reads the already-mutated target, so a device scatter that corrupts a non-indexed region is inherited by the reference and this `assert_close` still passes — the test can't catch that class of bug. Snapshot the reference inputs before the compiled run, or reuse `run_e2e(..., expect_close=True)` which already does. (The `index_copy` tests are functional, so unaffected.)

File: tests/inductor/test_indirect_access_scatter.py [hunk 2]
```diff
@@ -153,6 +164,129 @@ def kernel(out, src, idx):
...
+        (stick_count, elems_per_stick) -- mirroring the real paged-KV-cache
+        layout used by attention decode/prefill (see test_paged.py)."""
+        eps = get_elem_in_stick(torch.float16)
+        return SpyreTensorLayout(
+            device_size=[L, H, (D + eps - 1) // eps, eps],
+            stride_map=[H * D, D, eps, 1],
+            device_dtype=get_device_dtype(torch.float16),
+        )
+
+    def _paged_kv_cache_operands(self, L=576, H=8, D=128, P=3):
```
Comment (ani300) [inline tests/inductor/test_indirect_access_scatter.py:182]: These operands (and the 3D/4D tests) skip `self.name_dims`, unlike every other scenario in the file. Tests pass, so the path tolerates unnamed dims — but please confirm the paged tiling the docstrings describe actually engages without naming.

File: tests/inductor/test_indirect_access_scatter.py [hunk 3]
```diff
@@ -64,7 +69,13 @@ def _full_index_store(self, M=128, N=256, P=3, dtype=torch.int32):
         """Operands for scatter with a full [P,N] index tensor: out[M,N], src[P,N]."""
         out = torch.zeros(M, N, dtype=torch.float16).to("spyre")
         src = torch.rand(P, N, dtype=torch.float16).to("spyre")
-        index = torch.randint(0, M, (P, N), dtype=dtype).to("spyre")
+        index = (
```
Comment (ani300) [inline tests/inductor/test_indirect_access_scatter.py:75]: `randperm(M)[:P].expand(P,N)` is collision-free but row-uniform, so per-element index variation within a row is no longer exercised by any caller (`test_scatter`, `test_scatter_add`, `test_scatter_reduce`). If intentional, a one-line note would make the coverage trade explicit.

**Review by ani300 — COMMENTED** [review]
Approve with nits — the feature works and CI is green on device; nice to see indirect-output scatter passing e2e for the paged-cache shapes. A few non-blocking test-quality points below.

**1. Reference is built from already-mutated inputs (main nit).** In `_assert_compiled_matches_cpu`, the compiled kernel runs first and, for the five `y[idx] = src` tests, mutates the device target in place; the CPU reference is then computed from `.cpu()` of that already-mutated target. The consequence: a device scatter that corrupts a **non-indexed** region (over-broad write, wrong stride into neighbor rows, stale-buffer coherence) is inherited identically by the reference and `assert_close` passes — the tests can't catch that class of regression. A wrong value *at* the indexed rows is still caught; `index_copy` (functional) is unaffected. Fix: snapshot the reference from pristine inputs before the compiled run, or route through the existing `run_e2e(self, kernel, *dev_args, expect_close=True)` in `indirect_access_common.py`, which already does the pre-run snapshot plus backend-abort handling and a dynamo reset (and would remove the duplicate e2e runner).

**2. `name_dims` omitted.** Every other scenario in the file names its dims before compiling; the new paged-cache operands and the 3D/4D tests don't. They pass, so the path tolerates unnamed dims — but worth confirming the paged tiling the docstrings claim ("L tiled into stick groups") actually engages without naming, else the docstring oversells what's exercised.

**3. `_full_index_store` coverage narrows.** Moving from `randint(0,M,(P,N))` to `randperm(M)[:P].expand(P,N)` makes indices collision-free (good) but row-uniform, so per-element index variation within a row is no longer exercised by any caller. Fine if deliberate — a one-line note would make the trade explicit.

Minor: 1 commit behind main (which has since added `test_masked_scatter_unexpanded_row_broadcast`); trivial rebase.

**Review by dgrove-oss — DISMISSED** [review]
None.

SUGGESTIONs

- [Still open] The new _paged_kv_cache_operands helper and the test_index_put_3d_scatter_dim1_e2e/test_index_put_4d_scatter_dim1_e2e tests never call self.name_dims, unlike every other scenario in the file (verified: grep name_dims shows no hits for these new operands even after the follow-up commit). Tests pass, so the path tolerates unnamed dims, but the docstrings claim paged tiling ("L tiled into stick groups") is being exercised — worth the author confirming that's actually engaged without dim naming, or adding the naming calls for consistency with the rest of the file.
- (Already fixed by 37dcbdc3) Reference-snapshot-after-mutation: _assert_compiled_matches_cpu now clones/.cpu()s inputs before the compiled run, matching run_e2e's pattern.
- (Already addressed by 37dcbdc3) Row-uniform index coverage gap in _full_index_store now has an explanatory docstring note.

QUESTIONs

- Minor, from the existing review: the branch was 1 commit behind main at review time (main has since added test_masked_scatter_unexpanded_row_broadcast) — worth a trivial rebase before merge, unrelated to correctness.

Verdict

COMMENT (approve once the name_dims point is addressed or explicitly waived)

CI is fully green (139 jobs), DCO is signed off on both commits, license header intact, pre-commit (ruff/mypy/import-regex) passes clean. Two of ani300's three points are already resolved in the follow-up commit; the name_dims omission on the new paged-cache and 3D/4D tests is the one open item — not a correctness blocker since tests pass, but worth a reply from the author confirming the paged-tiling path is genuinely exercised, since that's the entire point these new tests were added for.

File: tests/inductor/test_indirect_access_scatter.py (hunk 1 above)
Comment (avery-blanchard) [reply tests/inductor/test_indirect_access_scatter.py:209]: Addressed in 37dcbdc397c73592edcb57064027426092d583df

File: tests/inductor/test_indirect_access_scatter.py (hunk 3 above)
Comment (avery-blanchard) [reply tests/inductor/test_indirect_access_scatter.py:75]: Comment added in 37dcbdc397c73592edcb57064027426092d583df

File: tests/inductor/test_indirect_access_scatter.py (hunk 2 above)
Comment (avery-blanchard) [reply tests/inductor/test_indirect_access_scatter.py:182]: Name dims are not used in the new tests. The comments refer to device layout tiling. I can clarify!

Comment (avery-blanchard) [reply tests/inductor/test_indirect_access_scatter.py:182]: Addressed in 7f57c96a

**Review by dgrove-oss — APPROVED** [review]
lgtm
